src/Elastic.Apm.SqlClient/SqlClientDiagnosticListener.cs

```diff
@@ -0,0 +1,174 @@
+using Elastic.Apm.Logging;
+
+namespace Elastic.Apm.SqlClient
+{
+	internal class SqlClientDiagnosticListener : IDiagnosticListener
+	{
+		private class PropertyFetcherSet
+		{
+			public PropertyFetcher StartCorrelationId { get; } = new PropertyFetcher("OperationId");
+			public PropertyFetcher StopCorrelationId { get; } = new PropertyFetcher("OperationId");
+			public PropertyFetcher ErrorCorrelationId { get; } = new PropertyFetcher("OperationId");
+
+			public PropertyFetcher Statistics { get; } = new PropertyFetcher("Statistics");
+
+			public PropertyFetcher CommandText { get; }
+			public PropertyFetcher CommandType { get; }
+			public PropertyFetcher Database { get; }
+
+			public PropertyFetcher Exception { get; } = new PropertyFetcher("Exception");
+
+			public PropertyFetcherSet()
+			{
+				var commandPropertyFetcher = new PropertyFetcher("Command");
+				var connectionPropertyFetcher = new CascadePropertyFetcher(commandPropertyFetcher, "Connection");
+
+				CommandText = new CascadePropertyFetcher(commandPropertyFetcher, "CommandText");
+				CommandType = new CascadePropertyFetcher(commandPropertyFetcher, "CommandType");
+
+				Database = new CascadePropertyFetcher(connectionPropertyFetcher, "Database");
+			}
+		}
+
+		private readonly IApmAgent _apmAgent;
+		private readonly IApmLogger _logger;
+
+		private readonly ConcurrentDictionary<Guid, ISpan> _processingQueries = new ConcurrentDictionary<Guid, ISpan>();
+
+		private readonly PropertyFetcherSet _systemPropertyFetcherSet = new PropertyFetcherSet();
+		private readonly PropertyFetcherSet _microsoftPropertyFetcherSet = new PropertyFetcherSet();
+
+
+		public SqlClientDiagnosticListener(IApmAgent apmAgent)
+		{
+			_apmAgent = apmAgent;
+			_logger = _apmAgent.Logger.Scoped(nameof(SqlClientDiagnosticListener));
+		}
+
+		public string Name => "SqlClientDiagnosticListener";
+
+		// prefix - Microsoft.Data.SqlClient. or System.Data.SqlClient.
+		public void OnNext(KeyValuePair<string, object> value)
+		{
+			if (value.Key.StartsWith("Microsoft.Data.SqlClient.") || value.Key.StartsWith("System.Data.SqlClient."))
+			{
+				switch (value.Key)
+				{
+					case string s when s.EndsWith("WriteCommandBefore") && _apmAgent.Tracer.CurrentTransaction != null:
+						HandleStartCommand(value.Value, value.Key.StartsWith("System") ? _systemPropertyFetcherSet : _microsoftPropertyFetcherSet);
+						break;
+					case string s when s.EndsWith("WriteCommandAfter"):
+						HandleStopCommand(value.Value, value.Key.StartsWith("System") ? _systemPropertyFetcherSet : _microsoftPropertyFetcherSet);
+						break;
+					case string s when s.EndsWith("WriteCommandError"):
+						HandleErrorCommand(value.Value, value.Key.StartsWith("System") ? _systemPropertyFetcherSet : _microsoftPropertyFetcherSet);
+						break;
+				}
+			}
+		}
+
+		private void HandleStartCommand(object payloadData, PropertyFetcherSet propertyFetcherSet)
+		{
+			try
+			{
+				var transaction = _apmAgent.Tracer.CurrentTransaction;
```

Comment:
It seems most of this code is almost the same as [code in `DbSpanCommon` ](https://github.com/elastic/apm-agent-dotnet/blob/1.2.0/src/Elastic.Apm/Model/DbSpanCommon.cs). Refactoring all this code into one location will simplify future features such as https://github.com/elastic/apm-agent-dotnet/issues/650 and https://github.com/elastic/apm-agent-dotnet/issues/611.

Reply:
I'll look into it

Reply:
@SergeyKleyman, I changed it to use `DbSpanCommon`. But I have a question, Do we have any reason to fill span info from db command during end operation instead of start (as it's done right now in db span common)? In case of exception, the end event isn't called and we don't capture an invalid db statement (if it's the reason), which would be nice to have.

Reply:
@SergeyKleyman, never mind. I don't see that you actually call `DbSpanCommon.End` from error  event

Reply:
I think capturing whatever possible in `Start` instead of waiting for `End` might be a good idea. I even considered if this change might affect unusual use cases such as [this one](https://discuss.elastic.co/t/get-currentspan-in-net-apm-agent/211741/5) and I think it shouldn't.